### DISCUSSION TITLE
When using Input mixer, how do I ensure a looped sound continues after additional sounds are finished?

### DISCUSSION THREAD

**Discussion by MJenkings** [pschatzmann/arduino-audio-tools#1500] [Q&A] [answered]
I am using the input mixer function to mix 2 wav streams from an SD card, it works perfectly. However, one file is 2 seconds long and must loop indefinitely while another is about 30 secs long and only plays once. The problem I have is that, when the 30 second file is finished, the short file just stops. I want it to continue looping but cannot figure out how to do this.  Any help would be appreciated.

**Comment by pschatzmann** [comment, marked as answer]
Just wrap the file in a [FileLoop](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_file_loop.html)! 
and use the file loop as data source.

**Comment by MJenkings** [comment]
Thanks for the assist. I have used the FileLoop as suggested but it still isn't looping indefinitely. The problem appears to be that when the longer file finishes, the ESP32-S3 just hangs hence why it isn't continuing the looping. Because it hangs, I get no error message. Only a hard power cycle brings it to life again. 

This is the entire code in the sketch 



#include "AudioTools.h"
#include "AudioLibs/AudioSourceSD.h"
#include "AudioLibs/AudioKit.h"
#include "Audiotools/AudioOutput.h"
#include "Audiotools/AudioIO.h"
#include "AudioLibs/FileLoop.h"
#include <SPI.h>
#include <SD.h>
#include "AudioCodecs/CodecWAV.h"

//config for sd card reader
#define SD_CS 5
#define SPI_MOSI 41
#define SPI_MISO 40
#define SPI_SCK  39


File audioFile1;

uint16_t sample_rate=44100;
InputMixer<int16_t> mixer;   

I2SStream i2s; // final output of decoded stream
MultiOutput multi;
VolumeStream vol(i2s); // sets volume and passes audio to i2s
StreamCopy sfx(multi, mixer); // copies sound into vol & out from mixer (destination , source)
EncodedAudioStream decoder(&vol, new MP3DecoderHelix()); // Decoding stream


VolumeOutput out;

void printMetaData(MetaDataType type, const char* str, int len){
  Serial.print("==> ");
  Serial.print(toStr(type));
  Serial.print(": ");
  Serial.println(str);
}

FileLoop loopingFile;

void setup() {


 Serial.begin(115200);
  decoder.setNotifyAudioChange(i2s);
  decoder.begin();
  multi.add(vol);
  multi.add(out);

  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // setup output

  
 auto config = i2s.defaultConfig(TX_MODE);
    
  config.pin_ws = 14;
  config.pin_bck = 13;
  config.pin_data = 11;
  config.sample_rate = sample_rate ;
  config.channels = 1;
  i2s.begin(config);
  out.begin(config);

SPI.begin(SPI_SCK, SPI_MISO, SPI_MOSI);

  SD.begin(SD_CS);

  vol.begin(config); // we need to provide the bits_per_sample and channels
  vol.setVolume(0.2);
   
  
  audioFile1 = SD.open("/audio/theme.wav");

  loopingFile.setFile(SD.open("/audio/scanner.wav"));
  loopingFile.setLoopCount(-1); // define loop count
  loopingFile.begin();

  mixer.add(audioFile1,25); //theme
  mixer.add(loopingFile,75);
  mixer.begin(config);

}

void loop() {
	sfx.copy();
	Serial.print("Volume: ");
	Serial.println(out.volume());

}

**Comment by pschatzmann** [comment]
Wouln't you need to wrap both files if you want to play endlessly ?

**Reply by MJenkings** [reply]
I only want one file to play on loop and the other to play one time. This works at the start, with the longer file (audioFile1) playing once while the other loops. However, once the longer file has finished playing once, everything stops and the ESP32 hangs.

**Reply by pschatzmann** [reply]
On the mixer try to call setLimitToAvailableData(true)

**Comment by MJenkings** [comment]
Good news; it fixed the hanging issue. The ESP32 no longer hangs. However, both audio streams stop when the longer one comes to an end. The looping one still doesn't continue to loop.

**Comment by pschatzmann** [comment]
This is up to you now: You need to remove the ended input out of the mixer or just stop using the mixer if you want to continue with the output of the loop

**Reply by pschatzmann** [reply]
I just tried [this example](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/streams-memory_pcm-mixer-audiokit/streams-memory_pcm-mixer-audiokit.ino) by commenting out the guitar loop, and I can't reproduce your issue (with the hanging).
Are you sure that you have the latest AudioTools version installed ?

**Reply by MJenkings** [reply]
I only installed it last month so I am pretty sure I have the latest version. However, to be sure I have uninstalled and reinstalled everything.  I was able to recreate your example and it does work, looping correctly and continuing to loop even after one has stopped. I can also substitute memory stream for Wav files from SD and it works fine. I can only assume that reinstalling everything helped fix this. However.  do I now have 3 new problems

1) If I call mixer.size() it always seems to return a value of 12 which doesn't seem to make sense, pretty sure this was returning correct values of 0,1 or 2 before. 
2) If I call mixer.setWeight() while the audio is running, it crashed the ESP32 -S3 and causes it to reboot, again, this was working before. 
3) If I add #include "AudioCodecs/CodecMP3Helix.h" to the sketch it will not compile at all. It gives no specific reason other than "Compilation error: exit status 1".  This seems to happen with the example sketches that include this file too.

**Reply by pschatzmann** [reply]
1. I committed a correction for the size() issue. 
2. Not sure what you were doing: but I added some additional check to prevent a divide by 0 error should somebody set the weigths all to 0. I could not reproduce your issue!
3. I guess you might have forgotten to install libhelix or have an outdated version

**Reply by MJenkings** [reply]
Thanks so much, I really appreciate your help. 

1) the size issue is now fixed - many thanks
2) Originally I was just setting the weight to 50 on a stream that was already 50, however this issue has also now disappeared since applying the update to AudioStreams.h. (Could it have been dependant on the size function?)
3) I had a libhelix library that was one version out of date, I had forgotten that this was dependant on that library - apologies for wasting your time on that. The update to this library has cured the compiling error. 

Do you have any plans to add a function to mixer to remove a stream? (the opposite to mixer.add(), it would be useful to be able to clear a stream completely so it can be replaced with another file. 

Many thanks again

**Reply by pschatzmann** [reply]
It is my expectation that calling setWeight() with a weight to 0 should be good enough to have a stream suppressed.
In which scenario would this be insufficient ?

**Reply by MJenkings** [reply]
I have a quantity of different sounds I may want to mix over the background sound. so I would like to be able to suppress an input on the mixer, remove it and replace it with a new sound without interrupting the looping input.

**Reply by pschatzmann** [reply]
I don't want to support the removal because this will mess up the index values and you will tend to loose the oversight which tone is on which channel (index) in order to update the volume correctly. 

However replacing is easy: I have added a [set() method](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_input_mixer.html#a3415dd8e48963708847d9e6c88afe34c) which let's you replace the stream at the indicated channel (index). 

I would expect, that it would also work. if you add all possible sounds with weight 0 and have just the ones active that you want to play....

**Reply by MJenkings** [reply]
Ok that sounds like it will do what I need then. Fantastic.  One final question.  How can I tell if the steam on a channel has come to the end of the file? For a non-looping sound, it would be useful to detect when the sound has finished playing.

**Reply by pschatzmann** [reply]
Usually just calling available() on the stream will tell you if there is still more data

**Reply by MJenkings** [reply]
I did try that in the mixer.available() and the Stream it feeds the mixer to, but they just continue to return a value of 1024 even after the mixer streams had ended.

**Reply by pschatzmann** [reply]
For loopingFile I would expect this (since it is endless), but not for [audioFile1](https://www.arduino.cc/reference/en/libraries/sd/available/): there I would expect that you should get the remaining bytes and 0 when it is at the end.

Alternatively you can get the file position and compare it with the file size.

**Reply by MJenkings** [reply]
Thank you, I have now resolved all this issues with this. It's all working perfectly.
